notebooks/desy1.py

```diff
@@ -0,0 +1,296 @@
+                   for i in range(1,5)]
+
+        self.nzs_l = [jc.redshift.kde_nz(nz_lens['Z_MID'].astype('float32'),
+                                      nz_lens['BIN%d'%i].astype('float32'), bw=0.01)
+                   for i in range(1,6)]    
+
+        # Define some ell range
+        self.ell = np.logspace(1, 3)
+        fid_cosmo = jc.Cosmology(sigma8=0.801,
+                                  Omega_c=0.2545,
+                                  Omega_b=0.0485,
+                                  h=0.682,
+                                  n_s=0.971,
+                                  w0=-1., Omega_k=0., wa=0.)
+        self.fid_params  = get_params_vec(fid_cosmo, 
+                                          [0., 0., 0., 0.],
+                                          [0., 0., 0., 0.],
+                                          [0.5, 0.],
+                                          [1.2, 1.4, 1.6, 1.8, 2.0])
+
+
+        # Set data as fiducial theory
+        print("Initializing mock likelihood")
+        self.args = [self.nzs_s, self.nzs_l, self.ell]
+        self.data_mean = theory_mean(self.fid_params, self.nzs_s, self.nzs_l, self.ell)
+        self.data_cov = theory_cov(self.fid_params, *self.args)
+        self.data_inv_cov = symmetric_positive_definite_inverse(self.data_cov)
+
+
+    def fisher_matrix(self, p0):
+        j = theory_jacobian(p0, *self.args)
+        j = onp.array(j)
+        F = onp.einsum('ia,ij,jb->ab', j, self.data_inv_cov, j)
+        F = 0.5*(F + F.T)
+
+        return F
+
+    def cov_estimate(self):
+        F = self.fisher_matrix(self.fid_params)
+        C = symmetric_positive_definite_inverse(F)
+        return C
+
+
+    def priors(self, p):
+        #priors - index, mean, std. dev
+        prior_values = [
+            (6, 0.012, 0.023),  #m1
+            (7, 0.012, 0.023),  #m2
+            (8, 0.012, 0.023),  #m3
+            (9, 0.012, 0.023),  #m4
+            (10, -0.001, 0.016),   #dz1
+            (11, -0.019, 0.013),   #dz2
+            (12,0.009, 0.011),   #dz3
+            (13, -0.018, 0.022),   #dz4
+        ]
+
+        logpi = 0.0
+        dlogpi_dp = onp.zeros_like(p)
+        for i, mu_i, sigma_i in prior_values:
+            logpi += -0.5 * (p[i] - mu_i)**2 / sigma_i**2
+            dlogpi_dp[i] = - (p[i] - mu_i) / sigma_i**2
+        return logpi, dlogpi_dp
+
+    def posterior(self, p):
+        cl = theory_mean(p, *self.args)
+        d = cl - self.data_mean
+        logL = -0.5 * d @ self.data_inv_cov @ d
+        logPi, _ = self.priors(p)
+        logP = logL + logPi
+        return logP
+
+    # returns -posterior P and -dL/dP
+    def posterior_and_gradient(self, p):
+        # theory C_ell prediction
+        cl = theory_mean(p, *self.args)
+        # d C_ell / d p
+        j = theory_jacobian(p, *self.args).T
+        d = cl - self.data_mean
+        dlogL_dCl = -self.data_inv_cov @ d
+        logL = 0.5 * d @ dlogL_dCl
+        dlogL_dp = j @ dlogL_dCl
```

Comment:
ah yeah, here we should be able to get some extra speed up, I *suspect* doing the jacobian computation separately here is going to be slower than the optimized `jax.value_and_grad` :
https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Evaluate-a-function-and-its-gradient-using-value_and_grad